### DISCUSSION TITLE
Correct way of marshalling strings

### DISCUSSION THREAD

**Discussion by Symbai** [dotnet/runtime#69347] [Q&A]
The official docs are just bouncing around each other, for example you go [here](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters) and it says for more information go [here](https://docs.microsoft.com/en-us/dotnet/framework/interop/default-marshalling-for-strings#fixed-length-string-buffers) which then also says for more information go to [first link]. Both however say: Avoid using `StringBuilder`. But there is one example given which uses `StringBuilder` 🤷‍♂️

I've [reported](https://github.com/dotnet/docs/issues/28315) to update the docs showing an example without StringBuilder, unfortunately it got "solved" by just updating the description but leaving the example untouched 🤷‍♂️

I've looked at the code in this repo but there are so many different versions all over the place. Some are using `StringBuilders`, others use `string` or `ValueStringBuilder` which is unfortunately not public.

Can someone provide a correct and performant example of what we should use nowadays? For example on `GetModuleFileName` WinAPI which second parameter returns the filename as `string`.

**Comment by MSDN-WhiteKnight** [comment]
[Native interoperability best practices](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters) has up-to-date recommendations: plain string marked as Charset.Unicode for readonly string, char array for read & write. Other pages you linked are in .NET Framework section, which means they are sort of maintainence-only and might be not updated to latest best practices. They might use StringBuilder for read & write string buffers, because it used to be common practice. You could try submitting PR to update them if you care, but for starters it would be good to figure out whether it really makes difference on .NET Framework (i.e. StringBuilder is significantly slower then char[]). It could be that the difference only comes from recent optimizations in .NET Core.

**Reply by Symbai** [reply]
The link you gave me is actually the first link in my post above. It does not give any examples, it just says try to avoid using Stringbuilder.

**Reply by danmoseley** [reply]
Cc @AaronRobinsonMSFT

**Reply by AaronRobinsonMSFT** [reply]
> Some are using StringBuilders, others use string or ValueStringBuilder which is unfortunately not public.

I don't think this is fair. The numbers here are vastly different relative to one another. I would be surprised if there are more than a dozen or so examples using `StringBuilder` in product P/Invoke signatures. The `ValueStringBuilder` type is, as mentioned, internal so not really an option. The correct way for `in` is to just use `string` - the first statement in the best-practices link. For `ref` and `out` the preference is, also on the best-practices page:

> ✔️ CONSIDER using char[] arrays from an ArrayPool.

A simple example for each case is likely warranted and removing the `StringBuilder` example is appropriate I think. I will submit a PR for the examples today.

**Reply by MSDN-WhiteKnight** [reply]
>The link you gave me is actually the first link in my post above.

I realize that. I didn't have better link to give. But i now see that analyzer page https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1838#how-to-fix-violations was linked in your docs issue and it have some examples. Indeed these examples should be added somewhere in interop docs.

>For ref and out the preference is, also on the best-practices page:  
>CONSIDER using char[] arrays from an ArrayPool.  

But not every app need to use ArrayPool. It could be regular array or stackalloced buffer. I think this recommendation could be expanded like this:

Use char[] arrays for out string parameters. To reduce memory pressure when you need to create many arrays, consider using stackalloc or ArrayPool.

**Reply by AaronRobinsonMSFT** [reply]
@Symbai I've tagged you and created a PR with updates - https://github.com/dotnet/docs/pull/29542. Please let us know if there are additional details missing that would be helpful.

**Comment by AaronRobinsonMSFT** [comment]
I recommend closing this and moving the conversation to https://github.com/dotnet/docs/pull/29542
